proposals/0000-custom-executors.md

```diff
@@ -0,0 +1,464 @@
+of the actor rather than requiring functions to suspend and resume
+executing asynchronously. This process is called *switching*. In
+situations where switching is impossible, such as when the actor is
+already executing on a thread, a job to process the actor will be
+scheduled onto the default concurrent executor.
+
+Switching significantly complicates the design of serial executors,
+which can be seen in the `SerialExecutor` protocol below. This
+complexity could be eliminated by only allowing default serial
+executors to participate in switching. Currently, we are reluctant
+to hardcode this limitation into the protocol design.
+
+### Actor executors
+
+An actor's executor must conform to `SerialExecutor`, which refines
+the `Executor` protocol. `SerialExecutor` makes additional guarantees
+about the behavior of `execute(currentExecutor:)` and also adds
+several new methods which are used to allow the executor to opt in
+to supporting "switching" serial executors the same way that the
+default serial executor does.
+
+TODO: if we need to support blocking actors, we will need to extend
+this significantly; otherwise, actor executors will need to assume
+that returning from a job should always unblock the executor.
+(Presumably we do not want to block actor executors by blocking
+their thread.)
+
+```swift
+protocol SerialExecutor: Executor {
+  /// Is it possible for this executor to give up the current thread
+  /// and allow it to start running a different actor?
+  var canGiveUpThread: Bool { get }
+
+  /// Given that canGiveUpThread() previously returned true, give up
+  /// the current thread.
+  func giveUpThread()
+
+  /// Attempt to start running a task on the current actor.  Returns
+  /// true if this succeeds.
+  func tryClaimThread() -> Bool
+}
+```
+
+All actors implicitly conform to the `Actor` protocol.  An actor
+must provide a serial executor.
+
+```swift
+protocol Actor: AnyObject {
+  /// Return the serial executoor for this actor.
+  ///
+  /// This must always return the same reference, and `isSerial()`
+  /// must return true for the reference.
+  ///
+  /// Keeping the actor reference valid must be sufficient to keep
+  /// the executor reference valid.
+  var serialExecutor: UnownedExecutorRef { get }
+}
+```
+
+An `actor` may derive its executor implementation in one
+of the following ways. We may add more ways in the future.
+
+- The actor may declare a property named `serialExecutor`.
+
+- The actor may declare a property named `delegateActor`.
+  The type of the property must be convertible to `Actor`.
+  The property must always return the same actor. Actor safety
+  checking should allow uses of the actor state and functions of
+  `delegateActor` from the actor functions of the delegating actor.
+
+  The `serialExecutor` property will be synthesized as if
+  the following:
+
+  ```swift
+  public final var serialExecutor: UnownedExecutorRef {
+    return delegateActor.serialExecutor
+  }
+  ```
```

Comment:
I said this before but here again: this is great 👍 

We already have a number of other people looking at APIs that will benefit from this 👍 (Funnily enough, same shapes as we had in our things before), it'll definitely be very useful.

That one might get some naming bikeshed (can't stop thinking about it as a form of `friend` hm...)  but let's do that on forums 😉